cylc/flow/scripts/workflow_state.py

```diff
@@ -18,252 +18,260 @@
-                my_point = my_parser.parse(self.args['cycle'], dump_format=fmt)
-                self.args['cycle'] = str(my_point)
-        return connected, self.args['cycle']
-
-    async def check(self):
-        """Return True if desired workflow state achieved, else False"""
-        return self.checker.task_state_met(
-            self.args['task'], self.args['cycle'],
-            self.args['status'], self.args['message'])
+    """A polling object that queries task states or outputs from a DB."""
+
+    def __init__(
+        self, id_, offset, flow_num, alt_cylc_run_dir, default_status,
+        is_output, old_format,
+        *args, **kwargs
+    ):
+        self.id_ = id_
+        self.offset = offset
+        self.flow_num = flow_num
+        self.alt_cylc_run_dir = alt_cylc_run_dir
+        self.old_format = old_format
+
+        self.db_checker = None
+
+        tokens = Tokens(self.id_)
+        self.workflow_id_raw = tokens.workflow_id
+        self.task_sel = tokens["task_sel"] or default_status
+        self.cycle_raw = tokens["cycle"]
+        self.task = tokens["task"]
+
+        self.workflow_id = None
+        self.cycle = None
+        self.results = None
+        self.db_checker = None
+
+        self.is_output = (
+            is_output or
+            (
+                self.task_sel is not None and
+                self.task_sel not in TASK_STATUSES_ORDERED
+            )
+        )
+
+        super().__init__(*args, **kwargs)
+
+    def _find_workflow(self) -> bool:
+        """Find workflow and infer run directory, return True if found."""
+        try:
+            self.workflow_id = infer_latest_run_from_id(
+                self.workflow_id_raw,
+                self.alt_cylc_run_dir
+            )
+        except InputError:
+            LOG.debug("Workflow not found")
+            return False
+
+        # Print inferred workflow ID.
+        sys.stderr.write(f"{self.workflow_id}\n")
+
+        return True
+
+    def _db_connect(self) -> bool:
+        """Connect to workflow DB, return True if connected."""
+        try:
+            self.db_checker = CylcWorkflowDBChecker(
+                get_cylc_run_dir(self.alt_cylc_run_dir),
+                self.workflow_id
+            )
+        except (OSError, sqlite3.Error):
+            LOG.debug("DB not connected")
+            return False
+
+        return True
+
+    async def check(self) -> bool:
+        """Return True if requested state achieved, else False.
+
+        Called once per poll by super() so only find and connect once.
+
+        Store self.result for external access.
+
+        """
+        if self.workflow_id is None and not self._find_workflow():
+            return False
+
+        if self.db_checker is None and not self._db_connect():
+            return False
+
+        if self.cycle is None:
+            # Adjust target cycle point to the DB format.
+            self.cycle = self.db_checker.adjust_point_to_db(
```

Comment:
Can we not do this once when we instantiate the object to avoid calling it each time?

Reply:
I considered that, but in principle you could query the connected object multiple times with different cycle points.